### DISCUSSION TITLE
Cloning broadcast receivers

### DISCUSSION THREAD

**Discussion by icrutt** [tokio-rs/tokio#4930] [Q&A] [answered]
I'm a Rust and Tokio newbie, so I apologise if this is a silly question.

I've been playing around with broadcast channels, and trying trying to find a way to connect spawned tasks together so they can send messages to one another. I find that I can clone a broadcast::Sender, but not a broadcast::Receiver. My understanding is that a broadcast channel can have multiple senders and receivers, so presumably it should be possible to clone both... What am I missing?

Here's my test code:
```
use tokio::sync::broadcast;

#[derive(Debug)]
struct MyStruct {
    tx: Option<broadcast::Sender<usize>>,
    rx: Option<broadcast::Receiver<usize>>,
}

impl MyStruct {
    fn new() -> Self {
        let (tx, mut rx) = broadcast::channel(16);
        Self {
            tx: Some(tx),
            rx: Some(rx),
        }
    }
}

#[tokio::main]
async fn main() {
    let my_struct = MyStruct::new();
    let tx = my_struct.tx.unwrap().clone();
    let rx = my_struct.rx.unwrap().clone();
}
```

**Comment by Darksonn** [comment, marked as answer]
The receiver provides a method called [`resubscribe`][1] that you can use for this. You can also call [`Sender::subscribe`][2] to create new receivers.

Making a clone that has the same position as the one you're cloning isn't currently possible. There has been some work on implementing it, but it causes some issues with the current way that the channel is implemented.

[1]: https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Receiver.html#method.resubscribe
[2]: https://docs.rs/tokio/latest/tokio/sync/broadcast/struct.Sender.html#method.subscribe

**Reply by icrutt** [reply]
Many thanks - that’s very helpful.
